GoCD agent on Java 20 can't talk to server on Java 17 - Exception parsing modifiedTime

##### Issue Type

- Bug Report

##### Summary

Hi team, 
I have recently updated my GoCD-Agent docker image using the latest softwares version and since then I have the problem that the GoCD-Agent can't parse the *modifiedTime* Git field and causes an exception. It is running into an Arch Linux Docker container (source below).

##### Environment

The GoCD server is running into a Docker container hosted on a QNAP NAS.

The GoCD-Agent causing the problem is running into a new Docker container hosted on a Windows 11 computer. However, a previous version of the Docker image of the GoCD-Agent runs without problem.

###### Basic environment details

* Go Version: `23.3.0 (16991-f38981e8aa684d63ac96f517996cd888eff37640)`
* JAVA Version: `17.0.8`
* OS: `Linux 5.10.60-qnap`

###### Additional Environment Details

_Dockerfile_ used to build the agent container image:

```
FROM archlinux:base-devel-20230723.0.166908

RUN pacman-key --init \
    && echo NoExtract = user/include/* >> /etc/pacman.conf \
    && echo NoExtract = user/share/* >> /etc/pacman.conf \
    && echo NoExtract = user/lib/*.a >> /etc/pacman.conf \
    && echo NoExtract = user/lib/libLLVM* >> /etc/pacman.conf \
    && echo NoExtract = user/lib/libgphobos* >> /etc/pacman.conf \
    && pacman -Syyu --noconfirm \
    && pacman -S --noconfirm \
        qt5-base \
        ffmpeg \
        glew \
        git \
        unzip \
        wget \
        eigen \
        hdf5 \
        libjpeg \
        libjpeg-turbo \
        libpng \
        libtiff \
        openssl \
        poco \
        freetype2 \
        pugixml \
        vtk \
        opencv \
        embree \
        libgit2 \
        jsoncpp \
        openmpi \
        fmt \
        vi \
        cmake \
        zip \
        go \
        gcc \
        gdb \
        gtest \
        jre-openjdk \
        extra/nodejs-lts-gallium \
        yarn \
        asciidoctor \
        extra/ansible-core \
        libpng \
        libjpeg-turbo \
        python-pip \
    && pacman --noconfirm -Sc \
    && rm -rf /var/cache/pacman/pkg

RUN yarn global add @quasar/cli node-gyp

RUN wget -O /tmp/go-agent.zip https://download.gocd.org/binaries/23.3.0-16991/generic/go-agent-23.3.0-16991.zip \
    && unzip -d /opt /tmp/go-agent.zip \
    && mv /opt/go-agent* /opt/go-agent \
    && rm /tmp/go-agent.zip \
    && sed -i "/#wrapper.lang=en_US/s/^#//g" /opt/go-agent/wrapper-config/wrapper.conf \
    && sed -i 's/localhost/gocd-server.local/g' /opt/go-agent/wrapper-config/wrapper-properties.conf

COPY entry.sh /entry.sh

CMD [ "/entry.sh" ]
```

The _entry.sh_ file content:

```
#!/bin/sh

rm -f /opt/go-agent/.agent-bootstrapper.running
/opt/go-agent/bin/go-agent console
```

##### Steps to Reproduce

1. Build the docker agent image from above source (replace the _gocd-server.local_)
2. Start the agent container (volumes on _/opt/go-agent/pipelines_, _/opt/go-agent/config_ and _/root/.ssh_ are mounted in my case)
3. Assign a pipeline with a Git material on the agent
4. Start a job

##### Actual Results

The agent cannot start the job and produces the logs below

##### Possible Fix

No workaround found until now:

* tried to set *LANG* variable to *en_US* in the container
* tried to force Java locale properties with _-Duser.country=US_ and _-Duser.language=en_

##### Log snippets

These logs are extracted from the _/opt/go-agent/logs/go-agent.log_:

```
2023-09-08 09:05:49,895 ERROR [scheduler-1] AgentController:99 - [Agent Loop] Error occurred during loop:
com.google.gson.JsonSyntaxException: Failed parsing 'Sep 2, 2023, 11:09:23 AM' as Date; at path $.assignment.materialRevisions.revisions[0].modifications[0].modifiedTime
        at com.google.gson.internal.bind.DateTypeAdapter.deserializeToDate(DateTypeAdapter.java:90)
        at com.google.gson.internal.bind.DateTypeAdapter.read(DateTypeAdapter.java:75)
        at com.google.gson.internal.bind.DateTypeAdapter.read(DateTypeAdapter.java:46)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.readIntoField(ReflectiveTypeAdapterFactory.java:212)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:433)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:393)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:40)
        at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
        at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.readIntoField(ReflectiveTypeAdapterFactory.java:212)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:433)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:393)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:40)
        at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
        at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.readIntoField(ReflectiveTypeAdapterFactory.java:212)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:433)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:393)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.readIntoField(ReflectiveTypeAdapterFactory.java:212)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:433)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:393)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.readIntoField(ReflectiveTypeAdapterFactory.java:212)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$FieldReflectionAdapter.readField(ReflectiveTypeAdapterFactory.java:433)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:393)
        at com.google.gson.TypeAdapter.fromJsonTree(TypeAdapter.java:299)
        at com.thoughtworks.go.remote.adapter.RuntimeTypeAdapterFactory$1.read(RuntimeTypeAdapterFactory.java:258)
        at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:204)
        at com.google.gson.Gson.fromJson(Gson.java:1227)
        at com.google.gson.Gson.fromJson(Gson.java:1137)
        at com.google.gson.Gson.fromJson(Gson.java:1047)
        at com.google.gson.Gson.fromJson(Gson.java:982)
        at com.thoughtworks.go.agent.RemotingClient.getWork(RemotingClient.java:77)
        at com.thoughtworks.go.agent.AgentHTTPClientController.doWork(AgentHTTPClientController.java:133)
        at com.thoughtworks.go.agent.AgentHTTPClientController.tryDoWork(AgentHTTPClientController.java:116)
        at com.thoughtworks.go.agent.AgentController.performWork(AgentController.java:89)
        at com.thoughtworks.go.agent.AgentWorkRetrievalScheduler.run(AgentWorkRetrievalScheduler.java:60)
        at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
        at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:81)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1623)
Caused by: java.text.ParseException: Failed to parse date ["Sep 2, 2023, 11:09:23 AM"]: Invalid number: Sep
        at com.google.gson.internal.bind.util.ISO8601Utils.parse(ISO8601Utils.java:279)
        at com.google.gson.internal.bind.DateTypeAdapter.deserializeToDate(DateTypeAdapter.java:88)
        ... 43 common frames omitted
Caused by: java.lang.NumberFormatException: Invalid number: Sep
        at com.google.gson.internal.bind.util.ISO8601Utils.parseInt(ISO8601Utils.java:316)
        at com.google.gson.internal.bind.util.ISO8601Utils.parse(ISO8601Utils.java:133)
        ... 44 common frames omitted
...
```

--- a/build.gradle
+++ b/build.gradle
@@ -274,6 +274,7 @@ subprojects {
     exclude '**/CommandLine.java'
     exclude '**/StreamPumper.java'
     exclude '**/StreamConsumer.java'
+    exclude '**/RuntimeTypeAdapterFactory*.java'
 
     // used by intellij for type hints when editing ftl files, intellij gets confused if it sees a license header in this file
     exclude '**/freemarker_implicit.ftl'

--- a/common/src/main/java/com/thoughtworks/go/remote/Serialization.java
+++ b/common/src/main/java/com/thoughtworks/go/remote/Serialization.java
@@ -17,6 +17,9 @@
 package com.thoughtworks.go.remote;
 
 import com.google.gson.*;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
 import com.thoughtworks.go.config.ArtifactStore;
 import com.thoughtworks.go.config.materials.PackageMaterial;
 import com.thoughtworks.go.config.materials.PluggableSCMMaterial;
@@ -48,10 +51,15 @@ import com.thoughtworks.go.server.service.AgentRuntimeInfo;
 import com.thoughtworks.go.server.service.ElasticAgentRuntimeInfo;
 
 import java.io.File;
+import java.io.IOException;
 import java.lang.reflect.Type;
 import java.nio.charset.Charset;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.Date;
 
 import static java.lang.String.format;
+import static java.time.format.DateTimeFormatter.ISO_INSTANT;
 import static org.apache.commons.io.FilenameUtils.separatorsToSystem;
 
 public class Serialization {
@@ -69,6 +77,7 @@ public class Serialization {
     @SuppressWarnings("deprecation")
     public static Gson create() {
         return new GsonBuilder()
+                .registerTypeAdapter(Date.class, new IsoInstantDateTypeAdapter())
                 .registerTypeAdapter(GoCipher.class, new SecurityRejectingAdapter<GoCipher>())
                 .registerTypeAdapter(AESEncrypter.class, new SecurityRejectingAdapter<AESEncrypter>())
                 .registerTypeAdapter(AESCipherProvider.class, new SecurityRejectingAdapter<AESCipherProvider>())
@@ -272,6 +281,31 @@ public class Serialization {
         @Override
         public String decrypt(String cipherText) throws CryptoException {
             throw new CryptoException("Agents should not be decrypting!");
+        }
+    }
+
+    private static class IsoInstantDateTypeAdapter extends TypeAdapter<Date> {
+        @Override
+        public void write(JsonWriter out, Date date) throws IOException {
+            if (date == null) {
+                out.nullValue();
+            } else {
+                out.value(ISO_INSTANT.format(date.toInstant()));
+            }
+        }
+
+        @Override
+        public Date read(JsonReader in) throws IOException {
+            if (in.peek() == JsonToken.NULL) {
+                in.nextNull();
+                return null;
+            }
+
+            try {
+                return new Date(ISO_INSTANT.parse(in.nextString(), Instant::from).toEpochMilli());
+            } catch (DateTimeParseException e) {
+                throw new JsonParseException(e);
+            }
         }
     }
 }

--- a/common/src/main/java/com/thoughtworks/go/remote/adapter/RuntimeTypeAdapterFactory.java
+++ b/common/src/main/java/com/thoughtworks/go/remote/adapter/RuntimeTypeAdapterFactory.java
@@ -1,20 +1,4 @@
 /*
- * Copyright 2023 Thoughtworks, Inc.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-/*
  * Copyright (C) 2011 Google Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -28,13 +12,11 @@
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
- */
-
-/*
- * This class has been copied from https://github.com/google/gson/blob/master/extras/src/main/java/com/google/gson/typeadapters/RuntimeTypeAdapterFactory.java
+ *
+ * Originally from Gson project. Modifications copyright Thoughtworks and respective GoCD contributors.
+ * This class has been copied from https://github.com/google/gson/blob/main/extras/src/main/java/com/google/gson/typeadapters/RuntimeTypeAdapterFactory.java
  * since the gson-extras library is not officially released or maintained, refer - https://github.com/google/gson/issues/845
-*/
-
+ */
 package com.thoughtworks.go.remote.adapter;
 
 import com.google.gson.*;
@@ -194,7 +176,7 @@ public final class RuntimeTypeAdapterFactory<T> implements TypeAdapterFactory {
      * sensitive.
      *
      * @throws IllegalArgumentException if either {@code type} or {@code label}
-     *     have already been registered on this type adapter.
+     *                                  have already been registered on this type adapter.
      */
     public RuntimeTypeAdapterFactory<T> registerSubtype(Class<? extends T> type, String label) {
         if (type == null || label == null) {
@@ -213,7 +195,7 @@ public final class RuntimeTypeAdapterFactory<T> implements TypeAdapterFactory {
      * name}. Labels are case sensitive.
      *
      * @throws IllegalArgumentException if either {@code type} or its simple name
-     *     have already been registered on this type adapter.
+     *                                  have already been registered on this type adapter.
      */
     public RuntimeTypeAdapterFactory<T> registerSubtype(Class<? extends T> type) {
         return registerSubtype(type, type.getSimpleName());
@@ -225,9 +207,9 @@ public final class RuntimeTypeAdapterFactory<T> implements TypeAdapterFactory {
         }
 
         final Map<String, TypeAdapter<?>> labelToDelegate
-                = new LinkedHashMap<>();
+            = new LinkedHashMap<>();
         final Map<Class<?>, TypeAdapter<?>> subtypeToDelegate
-                = new LinkedHashMap<>();
+            = new LinkedHashMap<>();
         for (Map.Entry<String, Class<?>> entry : labelToSubtype.entrySet()) {
             TypeAdapter<?> delegate = gson.getDelegateAdapter(this, TypeToken.get(entry.getValue()));
             labelToDelegate.put(entry.getKey(), delegate);
@@ -235,7 +217,8 @@ public final class RuntimeTypeAdapterFactory<T> implements TypeAdapterFactory {
         }
 
         return new TypeAdapter<R>() {
-            @Override public R read(JsonReader in) throws IOException {
+            @Override
+            public R read(JsonReader in) throws IOException {
                 JsonElement jsonElement = Streams.parse(in);
                 JsonElement labelJsonElement;
                 if (maintainType) {
@@ -246,26 +229,27 @@ public final class RuntimeTypeAdapterFactory<T> implements TypeAdapterFactory {
 
                 if (labelJsonElement == null) {
                     throw new JsonParseException("cannot deserialize " + baseType
-                            + " because it does not define a field named " + typeFieldName);
+                        + " because it does not define a field named " + typeFieldName);
                 }
                 String label = labelJsonElement.getAsString();
                 @SuppressWarnings("unchecked") // registration requires that subtype extends T
                 TypeAdapter<R> delegate = (TypeAdapter<R>) labelToDelegate.get(label);
                 if (delegate == null) {
                     throw new JsonParseException("cannot deserialize " + baseType + " subtype named "
-                            + label + "; did you forget to register a subtype?");
+                        + label + "; did you forget to register a subtype?");
                 }
                 return delegate.fromJsonTree(jsonElement);
             }
 
-            @Override public void write(JsonWriter out, R value) throws IOException {
+            @Override
+            public void write(JsonWriter out, R value) throws IOException {
                 Class<?> srcType = value.getClass();
                 String label = subtypeToLabel.get(srcType);
                 @SuppressWarnings("unchecked") // registration requires that subtype extends T
                 TypeAdapter<R> delegate = (TypeAdapter<R>) subtypeToDelegate.get(srcType);
                 if (delegate == null) {
                     throw new JsonParseException("cannot serialize " + srcType.getName()
-                            + "; did you forget to register a subtype?");
+                        + "; did you forget to register a subtype?");
                 }
                 JsonObject jsonObject = delegate.toJsonTree(value).getAsJsonObject();
 
@@ -278,7 +262,7 @@ public final class RuntimeTypeAdapterFactory<T> implements TypeAdapterFactory {
 
                 if (jsonObject.has(typeFieldName)) {
                     throw new JsonParseException("cannot serialize " + srcType.getName()
-                            + " because it already defines a field named " + typeFieldName);
+                        + " because it already defines a field named " + typeFieldName);
                 }
                 clone.add(typeFieldName, new JsonPrimitive(label));
 

--- a/common/src/test/java/com/thoughtworks/go/remote/SerializationTest.java
+++ b/common/src/test/java/com/thoughtworks/go/remote/SerializationTest.java
@@ -25,6 +25,7 @@ import com.thoughtworks.go.domain.config.EncryptedConfigurationValue;
 import com.thoughtworks.go.helper.ReversingEncrypter;
 import com.thoughtworks.go.security.*;
 import com.thoughtworks.go.util.SystemEnvironment;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
@@ -33,6 +34,10 @@ import java.io.File;
 import java.io.IOException;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
+import java.sql.Timestamp;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.Date;
 import java.util.List;
 import java.util.Map;
 
@@ -144,6 +149,7 @@ class SerializationTest {
     @Test
     void serializesCharsets() {
         assertThatJson(Serialization.instance().toJson(StandardCharsets.UTF_8)).isEqualTo("UTF-8");
+        //noinspection CharsetObjectCanBeUsed
         assertThatJson(Serialization.instance().toJson(Charset.forName("ascii"))).isEqualTo("US-ASCII");
     }
 
@@ -153,11 +159,36 @@ class SerializationTest {
         assertEquals(StandardCharsets.US_ASCII, Serialization.instance().fromJson("ascii", Charset.class));
     }
 
+    @Nested
+    public class DateTimes {
+        final Date TEST_TIME = new Date(ZonedDateTime.of(2023, 12, 13, 1, 2, 3, 4000000, ZoneId.of("UTC")).toInstant().toEpochMilli());
+
+        @Test
+        void serializesDatesDeterministically() {
+            assertEquals("null", Serialization.instance().toJson((Date)null));
+            assertEquals("\"1970-01-01T00:00:00Z\"", Serialization.instance().toJson(new Date(0)));
+            assertEquals("\"2023-12-13T01:02:03.004Z\"", Serialization.instance().toJson(TEST_TIME));
+        }
+        @Test
+        void serializesSqlTimeStampsDeterministically() {
+            assertEquals("null", Serialization.instance().toJson((Timestamp)null));
+            assertEquals("\"1970-01-01T00:00:00Z\"", Serialization.instance().toJson(new Timestamp(0)));
+            assertEquals("\"2023-12-13T01:02:03.004Z\"", Serialization.instance().toJson(new Timestamp(TEST_TIME.getTime())));
+        }
+
+        @Test
+        void deserializesDatesDeterministically() {
+            assertNull(Serialization.instance().fromJson("null", Date.class));
+            assertEquals(new Date(0), Serialization.instance().fromJson("\"1970-01-01T00:00:00Z\"", Date.class));
+            assertEquals(TEST_TIME, Serialization.instance().fromJson("\"2023-12-13T01:02:03.004Z\"", Date.class));
+        }
+    }
+
     @SuppressWarnings("unchecked")
     @Test
     void successfullySerializesConfigurationPropertyBecauseGoCipherIsHiddenFromSerialization() {
         assertDoesNotThrow(() -> {
-            final String json = Serialization.instance().toJson(new ConfigurationProperty(dumbcipher())
+            final String json = Serialization.instance().toJson(new ConfigurationProperty(dumbCipher())
                     .withKey("hello")
                     .withEncryptedValue("dlrow"));
             Map<String, String> actual = new Gson().fromJson(json, Map.class);
@@ -230,16 +261,16 @@ class SerializationTest {
 
     @SuppressWarnings("SameParameterValue")
     private ConfigurationProperty plainProperty(String name, String value) {
-        return new ConfigurationProperty(new ConfigurationKey(name), new ConfigurationValue(value), null, dumbcipher());
+        return new ConfigurationProperty(new ConfigurationKey(name), new ConfigurationValue(value), null, dumbCipher());
     }
 
-    private GoCipher dumbcipher() {
+    private GoCipher dumbCipher() {
         return new GoCipher(new ReversingEncrypter());
     }
 
     @SuppressWarnings("SameParameterValue")
     private ConfigurationProperty secretProperty(String name, String encryptedValue) {
-        return new ConfigurationProperty(new ConfigurationKey(name), null, new EncryptedConfigurationValue(encryptedValue), dumbcipher());
+        return new ConfigurationProperty(new ConfigurationKey(name), null, new EncryptedConfigurationValue(encryptedValue), dumbCipher());
     }
 
     private static class TempSystemEnvironment extends SystemEnvironment {

--- a/common/src/test/java/com/thoughtworks/go/remote/adapter/RuntimeTypeAdapterFactoryTest.java
+++ b/common/src/test/java/com/thoughtworks/go/remote/adapter/RuntimeTypeAdapterFactoryTest.java
@@ -1,0 +1,211 @@
+/*
+ * Copyright (C) 2011 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Originally from Gson project. Modifications copyright Thoughtworks and respective GoCD contributors.
+ * This class has been copied from https://github.com/google/gson/blob/main/extras/src/test/java/com/google/gson/typeadapters/RuntimeTypeAdapterFactoryTest.java
+ * since the gson-extras library is not officially released or maintained, refer - https://github.com/google/gson/issues/845
+ */
+
+package com.thoughtworks.go.remote.adapter;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonParseException;
+import com.google.gson.TypeAdapterFactory;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.*;
+
+public final class RuntimeTypeAdapterFactoryTest {
+
+    @Test
+    public void testRuntimeTypeAdapter() {
+        RuntimeTypeAdapterFactory<BillingInstrument> rta = RuntimeTypeAdapterFactory.of(
+                BillingInstrument.class)
+            .registerSubtype(CreditCard.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(rta)
+            .create();
+
+        CreditCard original = new CreditCard("Jesse", 234);
+        assertEquals("{\"type\":\"CreditCard\",\"cvv\":234,\"ownerName\":\"Jesse\"}",
+            gson.toJson(original, BillingInstrument.class));
+        BillingInstrument deserialized = gson.fromJson(
+            "{type:'CreditCard',cvv:234,ownerName:'Jesse'}", BillingInstrument.class);
+        assertEquals("Jesse", deserialized.ownerName);
+        assertTrue(deserialized instanceof CreditCard);
+    }
+
+    @Test
+    public void testRuntimeTypeIsBaseType() {
+        TypeAdapterFactory rta = RuntimeTypeAdapterFactory.of(
+                BillingInstrument.class)
+            .registerSubtype(BillingInstrument.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(rta)
+            .create();
+
+        BillingInstrument original = new BillingInstrument("Jesse");
+        assertEquals("{\"type\":\"BillingInstrument\",\"ownerName\":\"Jesse\"}",
+            gson.toJson(original, BillingInstrument.class));
+        BillingInstrument deserialized = gson.fromJson(
+            "{type:'BillingInstrument',ownerName:'Jesse'}", BillingInstrument.class);
+        assertEquals("Jesse", deserialized.ownerName);
+    }
+
+    @Test
+    public void testNullBaseType() {
+        assertThatThrownBy(() -> RuntimeTypeAdapterFactory.of(null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void testNullTypeFieldName() {
+        assertThatThrownBy(() -> RuntimeTypeAdapterFactory.of(BillingInstrument.class, null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void testNullSubtype() {
+        RuntimeTypeAdapterFactory<BillingInstrument> rta = RuntimeTypeAdapterFactory.of(
+            BillingInstrument.class);
+        assertThatThrownBy(() -> rta.registerSubtype(null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void testNullLabel() {
+        RuntimeTypeAdapterFactory<BillingInstrument> rta = RuntimeTypeAdapterFactory.of(
+            BillingInstrument.class);
+        assertThatThrownBy(() -> rta.registerSubtype(CreditCard.class, null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void testDuplicateSubtype() {
+        RuntimeTypeAdapterFactory<BillingInstrument> rta = RuntimeTypeAdapterFactory.of(
+            BillingInstrument.class);
+        rta.registerSubtype(CreditCard.class, "CC");
+        assertThatThrownBy(() -> rta.registerSubtype(CreditCard.class, "Visa"))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessageContaining("types and labels must be unique");
+    }
+
+    @Test
+    public void testDuplicateLabel() {
+        RuntimeTypeAdapterFactory<BillingInstrument> rta = RuntimeTypeAdapterFactory.of(
+            BillingInstrument.class);
+        rta.registerSubtype(CreditCard.class, "CC");
+        assertThatThrownBy(() -> rta.registerSubtype(BankTransfer.class, "CC"))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessageContaining("types and labels must be unique");
+    }
+
+    @Test
+    public void testDeserializeMissingTypeField() {
+        TypeAdapterFactory billingAdapter = RuntimeTypeAdapterFactory.of(BillingInstrument.class)
+            .registerSubtype(CreditCard.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(billingAdapter)
+            .create();
+        assertThatThrownBy(() -> gson.fromJson("{ownerName:'Jesse'}", BillingInstrument.class))
+            .isInstanceOf(JsonParseException.class)
+            .hasMessageContaining("type");
+    }
+
+    @Test
+    public void testDeserializeMissingSubtype() {
+        TypeAdapterFactory billingAdapter = RuntimeTypeAdapterFactory.of(BillingInstrument.class)
+            .registerSubtype(BankTransfer.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(billingAdapter)
+            .create();
+        assertThatThrownBy(() -> gson.fromJson("{type:'CreditCard',ownerName:'Jesse'}", BillingInstrument.class))
+            .isInstanceOf(JsonParseException.class)
+            .hasMessageContaining("CreditCard");
+    }
+
+    @Test
+    public void testSerializeMissingSubtype() {
+        TypeAdapterFactory billingAdapter = RuntimeTypeAdapterFactory.of(BillingInstrument.class)
+            .registerSubtype(BankTransfer.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(billingAdapter)
+            .create();
+        assertThatThrownBy(() -> gson.toJson(new CreditCard("Jesse", 456), BillingInstrument.class))
+            .isInstanceOf(JsonParseException.class)
+            .hasMessageContaining("type");
+    }
+
+    @Test
+    public void testSerializeCollidingTypeFieldName() {
+        TypeAdapterFactory billingAdapter = RuntimeTypeAdapterFactory.of(BillingInstrument.class, "cvv")
+            .registerSubtype(CreditCard.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(billingAdapter)
+            .create();
+        assertThatThrownBy(() -> gson.toJson(new CreditCard("Jesse", 456), BillingInstrument.class))
+            .isInstanceOf(JsonParseException.class)
+            .hasMessageContaining("already defines a field named cvv");
+    }
+
+    @Test
+    public void testSerializeWrappedNullValue() {
+        TypeAdapterFactory billingAdapter = RuntimeTypeAdapterFactory.of(BillingInstrument.class)
+            .registerSubtype(CreditCard.class)
+            .registerSubtype(BankTransfer.class);
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(billingAdapter)
+            .create();
+        String serialized = gson.toJson(new BillingInstrumentWrapper(null), BillingInstrumentWrapper.class);
+        BillingInstrumentWrapper deserialized = gson.fromJson(serialized, BillingInstrumentWrapper.class);
+        assertNull(deserialized.instrument);
+    }
+
+    static class BillingInstrumentWrapper {
+        BillingInstrument instrument;
+
+        BillingInstrumentWrapper(BillingInstrument instrument) {
+            this.instrument = instrument;
+        }
+    }
+
+    static class BillingInstrument {
+        private final String ownerName;
+
+        BillingInstrument(String ownerName) {
+            this.ownerName = ownerName;
+        }
+    }
+
+    static class CreditCard extends BillingInstrument {
+        int cvv;
+
+        CreditCard(String ownerName, int cvv) {
+            super(ownerName);
+            this.cvv = cvv;
+        }
+    }
+
+    static class BankTransfer extends BillingInstrument {
+        int bankAccount;
+
+        BankTransfer(String ownerName, int bankAccount) {
+            super(ownerName);
+            this.bankAccount = bankAccount;
+        }
+    }
+}

